internal/pkg/upload/upload.go

```diff
@@ -0,0 +1,340 @@
+	Source    string // which integration is performing the upload
+	ChunkSize int64
+	Total     int64
+	Count     int
+	HashSum   string
+	Hasher    hash.Hash
+}
+
+type ChunkInfo struct {
+	ID            int
+	FirstReceived bool
+	Final         bool
+	Upload        Info
+	Token         *throttle.Token
+}
+
+func New(sizeLimit int64, opLimit int, chunkLimit int) *Uploader {
+	return &Uploader{
+		parallelChunkLimit: chunkLimit,
+		sizeLimit:          sizeLimit,
+		opThrottle:         throttle.NewThrottle(opLimit),
+		current:            make(map[string]upload, opLimit),
+	}
+}
+
+// Start an upload operation, as long as the max concurrent has not been reached
+// returns the upload ID
+func (u *Uploader) Begin(size int64, docID string, source string, hashsum string, hasher hash.Hash) (Info, error) {
+	if size <= 0 {
+		return Info{}, errors.New("invalid file size")
+	}
+	if size > u.sizeLimit {
+		return Info{}, ErrFileSizeTooLarge
+	}
+
+	uid, err := uuid.NewV4()
+	if err != nil {
+		return Info{}, fmt.Errorf("unable to generate upload operation ID: %w", err)
+	}
+	id := uid.String()
+
+	token := u.opThrottle.Acquire(id, 300*time.Hour)
+	if token == nil {
+		return Info{}, ErrMaxConcurrentUploads
+	}
+
+	total := time.NewTimer(uploadRequestTimeout)
+	chunkT := time.NewTimer(chunkProgressTimeout)
+	chunkRecv := make(chan struct{})
+	complete := make(chan struct{})
+	// total timer could also be achieved with context deadline and cancelling
+
+	go func() {
+		for {
+			select {
+			case <-total.C: // entire upload operation timed out
+				log.Trace().Str("uploadID", id).Msg("upload operation timed out")
+				// stop and drain chunk timer
+				if !chunkT.Stop() {
+					<-chunkT.C
+				}
+				u.cancel(id)
+				return
+			case <-chunkT.C: // no chunk progress within chunk timer, expire operation
+				log.Trace().Str("uploadID", id).Msg("upload operation chunk activity timed out")
+				// stop and drain total timer
+				if !total.Stop() {
+					<-total.C
+				}
+				u.cancel(id)
+				return
+			case <-chunkRecv: // chunk activity, update chunk timer
+				if !chunkT.Stop() {
+					<-chunkT.C
+				}
+				chunkT.Reset(chunkProgressTimeout)
+			case <-complete: // upload operation complete, clean up
+				if !chunkT.Stop() {
+					<-chunkT.C
+				}
+				if !total.Stop() {
+					<-total.C
+				}
+				u.finalize(id)
+				return
+			}
+		}
+	}()
+	info := Info{
+		ID:        id,
+		DocID:     docID,
+		ChunkSize: MaxChunkSize,
+		Source:    source,
+		Total:     size,
+		Hasher:    hasher,
+		HashSum:   hashsum,
+	}
+	cnt := info.Total / info.ChunkSize
+	if info.Total%info.ChunkSize > 0 {
+		cnt += 1
+	}
+	info.Count = int(cnt)
+	u.current[id] = upload{
+		opToken:       token,
+		chunkThrottle: throttle.NewThrottle(u.parallelChunkLimit),
+		complete:      complete,
+		chunkRecv:     chunkRecv,
+		Info:          info,
+	}
+	return info, nil
+}
+
+func (u *Uploader) Chunk(uplID string, chunkID int) (ChunkInfo, error) {
+	u.mu.Lock()
+	defer u.mu.Unlock()
```

Comment:
With this locking, 2 two chunks can be uploaded at once, is that desired?

Reply:
this function isn't performing the chunk upload, but checking that that chunk number (`chunkID`) _can_ be uploaded for that upload ID. 

Checking bounds (valid Chunk ID, and it is not an unexpectedly large ID for the given file size), and checks against the throttle for that upload.

A given upload can have a few chunks in parallel (the `chunkThrottle` below). Each upload has its own throttle.

This also returns some info so the uploading logic knows whether this is the final chunk in the sequence (which affects a parameter in the chunk document). It also is watching for the "first" chunk activity to start a total-upload timer.

It's probably more of "canChunkUploadProceed", but I expect some refactoring here anyway, as you pointed out the `dl` layer and things. I have no problem moving around _where_ the uploading takes place, or where some of these checks and responsibilities end up